src/jit/gentree.h

```diff
@@ -4626,24 +4636,58 @@ struct GenTreeAllocObj final : public GenTreeUnOp
 // be defined already.
 //------------------------------------------------------------------------
 
-inline bool GenTree::OperIsBlkOp() const
+inline bool GenTree::OperIsBlkOp()
 {
-    return (gtOper == GT_INITBLK || gtOper == GT_COPYBLK || gtOper == GT_COPYOBJ);
+    return (((gtOper == GT_ASG) && varTypeIsStruct(gtOp.gtOp1))
+#ifndef LEGACY_BACKEND
+            || (OperIsBlk() && (AsBlk()->Data() != nullptr))
+#endif
+                );
 }
 
 inline bool GenTree::OperIsDynBlkOp()
 {
-    return (OperIsBlkOp() && !gtGetOp2()->IsCnsIntOrI());
+    if (gtOper == GT_ASG)
+    {
+        return gtGetOp1()->OperGet() == GT_DYN_BLK;
+    }
+#ifndef LEGACY_BACKEND
+    else if (gtOper == GT_STORE_DYN_BLK)
+    {
+        return true;
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsCopyBlkOp() const
+inline bool GenTree::OperIsCopyBlkOp()
 {
-    return (gtOper == GT_COPYOBJ || gtOper == GT_COPYBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsInitBlkOp() const
+inline bool GenTree::OperIsInitBlkOp()
 {
-    return (gtOper == GT_INITBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) != 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) != 0));
```

Comment:
No because we can, in future, have block nodes for which the types have been modified to a fixed-size type, and the rhs may be a constant. The semantics of initBlk are somewhat unique in that it's not really an assignment of the full rhs, but rather an assignment of the lower 8 bits of the rhs constant expanded to fit the size of the block. I previously did not have the GTK_BLK_INIT flag, but it became messy to check for the cases, so this seems the better solution.

---
In reply to: [77381784](https://github.com/dotnet/coreclr/pull/7027#discussion_r77381784) [](ancestors = 77381784)